relay-server/src/services/buffer/envelope_repository/sqlite.rs

```diff
@@ -0,0 +1,645 @@
+        for &project_key_pair in project_key_pairs {
+            let envelope_stack = self.envelope_stacks.entry(project_key_pair).or_default();
+            // When creating an envelope stack during initialization, we assume data is on disk.
+            envelope_stack.check_disk = true;
+        }
+    }
+
+    /// Determines if the number of buffered envelopes is above the spool threshold.
+    fn above_spool_threshold(&self) -> bool {
+        self.buffered_envelopes_size >= self.disk_batch_size as u64
+    }
+
+    /// Spools all buffered envelopes to disk.
+    async fn spool_to_disk(&mut self) -> Result<(), SqliteEnvelopeRepositoryError> {
+        let batches = self.get_envelope_batches();
+        // We should have only one batch here, since we spool when we reach the batch size.
+        debug_assert!(batches.len() == 1);
+
+        for batch in batches {
+            self.insert_envelope_batch(batch).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Unspools from disk up to `n` envelopes and returns them.
+    async fn unspool_from_disk(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        n: u64,
+    ) -> Result<Vec<Box<Envelope>>, SqliteEnvelopeRepositoryError> {
+        let envelopes = relay_statsd::metric!(timer(RelayTimers::BufferUnspool), {
+            self.envelope_store
+                .delete_many(
+                    project_key_pair.own_key,
+                    project_key_pair.sampling_key,
+                    n as i64,
+                )
+                .await
+                .map_err(SqliteEnvelopeRepositoryError::EnvelopeStoreError)?
+        });
+
+        if envelopes.is_empty() {
+            // In case no envelopes were unspooled, we mark this project key pair as having no
+            // envelopes on disk.
+            self.set_check_disk(project_key_pair, false);
+
+            return Ok(vec![]);
+        }
+
+        relay_statsd::metric!(
+            counter(RelayCounters::BufferUnspooledEnvelopes) += envelopes.len() as u64
+        );
+
+        Ok(envelopes)
+    }
+
+    /// Returns `true` whether the disk should be checked for data for a given [`ProjectKeyPair`],
+    /// false otherwise.
+    fn should_check_disk(&self, project_key_pair: ProjectKeyPair) -> bool {
+        // If a project key pair is unknown, we don't want to check disk.
+        self.envelope_stacks
+            .get(&project_key_pair)
+            .map_or(false, |e| e.check_disk)
+    }
+
+    /// Sets on the [`EnvelopeStack`] whether the disk should be checked or not.
+    fn set_check_disk(&mut self, project_key_pair: ProjectKeyPair, check_disk: bool) {
+        if let Some(envelope_stack) = self.envelope_stacks.get_mut(&project_key_pair) {
+            envelope_stack.check_disk = check_disk;
+        }
+    }
+
+    /// Returns batches of envelopes of size `self.disk_batch_size`.
+    #[allow(clippy::vec_box)]
+    fn get_envelope_batches(&mut self) -> Vec<Vec<Box<Envelope>>> {
```

Comment:
Can we make this return an `impl Iterator`? In the `flush` case this vector can become pretty large.